proof/proof.go

```diff
@@ -0,0 +1,133 @@
+package proof
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ipfs/go-cid"
+	hamt "github.com/ipfs/go-hamt-ipld"
+	"github.com/opentracing/opentracing-go"
+	"github.com/quorumcontrol/chaintree/nodestore"
+	"github.com/quorumcontrol/messages/v2/build/go/gossip"
+	"github.com/quorumcontrol/tupelo-go-sdk/bls"
+	"github.com/quorumcontrol/tupelo-go-sdk/gossip/hamtwrapper"
+	sigfuncs "github.com/quorumcontrol/tupelo-go-sdk/signatures"
+)
+
+func verifyRoundConfirmation(rootCtx context.Context, conf *gossip.RoundConfirmation, quorumCount uint64, verKeys []*bls.VerKey) (bool, error) {
+	sp, ctx := opentracing.StartSpanFromContext(rootCtx, "verifyRoundConfirmation")
+	defer sp.Finish()
+
+	sig := conf.Signature
+	if uint64(sigfuncs.SignerCount(sig)) < quorumCount {
+		return false, nil
+	}
+
+	err := sigfuncs.RestoreBLSPublicKey(ctx, sig, verKeys)
+	if err != nil {
+		sp.SetTag("error", true)
+		return false, fmt.Errorf("error restoring public key: %v", err)
+	}
+
+	msg := conf.RoundCid
+	isVerified, err := sigfuncs.Valid(ctx, sig, msg, nil)
+	if err != nil {
+		sp.SetTag("error", true)
+		return false, fmt.Errorf("error verifying round signature: %v", err)
+	}
+	sp.SetTag("verified", isVerified)
+	return isVerified, nil
+}
+
+func Verify(rootCtx context.Context, proof *gossip.Proof, quorumCount uint64, verKeys []*bls.VerKey) (bool, error) {
+	sp, ctx := opentracing.StartSpanFromContext(rootCtx, "validateProof")
+	defer sp.Finish()
+
+	validRoundConfirmation, err := verifyRoundConfirmation(ctx, proof.RoundConfirmation, quorumCount, verKeys)
+	if err != nil {
+		return false, fmt.Errorf("error verifying round confirmation: %v", err)
+	}
+
+	if !validRoundConfirmation {
+		return false, nil
+	}
+
+	underlyingStore := nodestore.MustMemoryStore(ctx)
+	hamtStore := hamt.CborIpldStore{
+		Blocks: hamtwrapper.NewStore(underlyingStore),
+	}
+
+	roundCid, err := hamtStore.Put(ctx, proof.Round)
```

Comment:
noting that I think these CID calculations are incorrect vs what the client is doing.

Reply:
Well, that explains why the round and round confirmation cids aren't matching up in my tests!

Reply:
https://github.com/quorumcontrol/tupelo-go-sdk/pull/177/files#diff-67701a72352c0e31f086a1732e111565R15-R21